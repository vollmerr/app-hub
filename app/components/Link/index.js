/**
*
* Link
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { Link as RouterLink } from 'react-router-dom';
import styled from 'styled-components';

export const StyledLink = styled(RouterLink) `
  text-decoration: none;
`;

export const A = styled.a`
  text-decoration: none;
`;

class Link extends React.PureComponent {
  render() {
    const { to, href, children, ...props } = this.props;

    if (to) {
      return (
        <StyledLink to={to} {...props}>
          {children}
        </StyledLink>
      );
    }

    return (
      <A href={href} target={'_blank'} rel={'noopener noreferrer'} {...props}>
        {children}
      </A>
    );
  }
}

Link.propTypes = {
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Link;
