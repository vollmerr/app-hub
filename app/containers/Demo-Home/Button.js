import React from 'react';

const Button = ({ onClick, disabled, text }) => ( // eslint-disable-line
  <button
    onClick={onClick}
    disabled={disabled}
    style={{ background: disabled ? '#666' : '#333', color: '#fff', padding: '10px 15px', cursor: 'pointer' }}
  >{text}</button>
);

export default Button;
