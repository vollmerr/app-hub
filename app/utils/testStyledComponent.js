import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

export default testStyledComponent;

