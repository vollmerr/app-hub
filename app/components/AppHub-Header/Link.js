import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { CommandButton } from 'office-ui-fabric-react/lib/Button';

import StyledLink from 'components/Link';
import theme from 'utils/theme';

import Item from './Item';

export const ButtonStyled = styled(CommandButton) `
  color: ${theme.white};
  display: flex;
  justify-content: center;
  vertical-align: middle;
  border-width: 0;
  padding: 0 ${(props) => props.padding || '0'};
  min-width: ${theme.hub.headerHeight};
  height: ${theme.hub.headerHeight};
  box-sizing: border-box;
  text-align: center;
  line-height: normal;
  overflow: visible;
  transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  pointer-events: all;
  font-size: ${(props) => props.size || '24px'};

  &:hover,
  i.ms-Icon,
  &:hover i.ms-Icon,
  &.is-checked i.ms-Icon {
    color: ${theme.white};
  }
`;


export function Button({ text, children, onClick, panel, ...props }) { //eslint-disable-line
  const newOnClick = onClick ? () => onClick(panel) : null;

  return (
    <ButtonStyled
      onClick={newOnClick}
      {...props}
    >
      {text}
      {children}
    </ButtonStyled>
  );
}


class Link extends React.PureComponent {
  render() {
    const {
      iconProps,
      onClick,
      to,
      href,
      dark,
      title,
      checked,
      ...props
    } = this.props;

    const isLink = !!(onClick || to || href);
    const newIconProps = iconProps && { style: { fontSize: '20px' }, ...iconProps };

    const itemProps = {
      isLink,
      dark,
      checked,
    };

    const buttonProps = {
      dark,
      title,
      checked,
      onClick,
      ariaLabel: title,
      iconProps: newIconProps,
      ...props,
    };

    // handle routing
    if (to || href) {
      return (
        <Item {...itemProps}>
          <StyledLink to={to} href={href} >
            <Button {...buttonProps} />
          </StyledLink>
        </Item>
      );
    }

    // no links
    return (
      <Item {...itemProps}>
        <Button {...buttonProps} />
      </Item>
    );
  }
}

Link.propTypes = {
  panel: PropTypes.string,
  title: PropTypes.string,
  iconProps: PropTypes.object,
  onClick: PropTypes.func,
  text: PropTypes.string,
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
  dark: PropTypes.bool,
  checked: PropTypes.bool,
  padding: PropTypes.string,
};

export default Link;
