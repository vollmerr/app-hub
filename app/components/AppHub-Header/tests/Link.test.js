import React from 'react';
import { shallow } from 'enzyme';
import { CommandButton } from 'office-ui-fabric-react/lib/Button';
import testStyledComponent from 'utils/testStyledComponent';

import StyledLink from 'components/Link';
import Link, { ButtonStyled, Button } from '../Link';

testStyledComponent(ButtonStyled, CommandButton);

const href = 'http://www.google.com';
const to = '/test-route';


describe('<Link />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Link />);
  });

  it('should render a link by default', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(StyledLink).length).toEqual(0);
  });

  it('should render an internal link if passed `to`', () => {
    wrapper.setProps({ to });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(StyledLink).length).toEqual(1);
    expect(wrapper.find(StyledLink).prop('to')).toEqual(to);
  });

  it('should render an external link if passed `href`', () => {
    wrapper.setProps({ href });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(StyledLink).length).toEqual(1);
    expect(wrapper.find(StyledLink).prop('href')).toEqual(href);
  });

  it('should be a button', () => {
    expect(wrapper.find(Button).length).toEqual(1);
    wrapper.setProps({ href });
    expect(wrapper.find(Button).length).toEqual(1);
    wrapper.setProps({ to });
    expect(wrapper.find(Button).length).toEqual(1);
  });

  it('should generate the correct iconProps', () => {
    const iconProps = { name: 'test icon name' };
    const expected = { style: { fontSize: '20px' }, ...iconProps };
    wrapper.setProps({ iconProps });
    expect(wrapper.find(Button).prop('iconProps')).toEqual(expected);
  });
});


const onClick = jest.fn();
const panel = 'test panel';


describe('<Button />', () => {
  let wrapper;
  beforeEach(() => {
    onClick.mockReset();
    wrapper = shallow(<Button onClick={onClick} panel={panel} />);
  });

  it('should not have an onClick if not passed one', () => {
    wrapper.setProps({ onClick: undefined });
    wrapper.simulate('click');
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should pass the panel to the onClick', () => {
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalledWith(panel);
  });
});
