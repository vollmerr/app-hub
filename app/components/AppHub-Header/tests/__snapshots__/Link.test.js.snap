// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should render a link by default 1`] = `
<Item__Item
  isLink={false}
>
  <Button />
</Item__Item>
`;

exports[`<Link /> should render an external link if passed \`href\` 1`] = `
<Item__Item
  isLink={true}
>
  <Link
    href="http://www.google.com"
  >
    <Button />
  </Link>
</Item__Item>
`;

exports[`<Link /> should render an internal link if passed \`to\` 1`] = `
<Item__Item
  isLink={true}
>
  <Link
    to="/test-route"
  >
    <Button />
  </Link>
</Item__Item>
`;

exports[`<Link__ButtonStyled /> should render correctly 1`] = `
.c0 {
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
  border-width: 0;
  padding: 0 0;
  min-width: 50px;
  height: 50px;
  box-sizing: border-box;
  -webkit-text-align: center;
  text-align: center;
  line-height: normal;
  overflow: visible;
  -webkit-transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  pointer-events: all;
  font-size: 24px;
}

.c0:hover,
.c0 i.ms-Icon,
.c0:hover i.ms-Icon,
.c0.is-checked i.ms-Icon {
  color: #ffffff;
}

<CustomizedActionButton
  className="c0"
/>
`;
